### PR TITLE
feat(atsu): add PDC/CPDLC communication via Hoppie

### PR DESCRIPTION
**Pull request by svengcz**
## Summary of Changes
This PR adds PDC/CPDLC communication via the Hoppie.nl-interface to the A32NX.
It uses the MCDU and DCDU for the communication with the controllers.
The configuration is handled via the Settings-page of the EFB.

The CPDLC ring tone and the ATC button next to the master warning are added.

It integrates the full PDC communication and simple CPDLC instructions.
Instructions that require generated or modified answers or requests are out-of-scope for this changes.

Implemented ATC instructions:
- Logon/Logoff
- Handoff
- Station identifier
- Climb FLxxx, Fly heading xxx, Reduce/Increase speed xxx
- PDC

Introduced an ATSU manager with message formats for all AOC messages and parts of the ATC messages.
Rewrote ATSU/AOC pages to use a central ATSU manager and its types.
Introduced a central datalink manager to synchronize the messages and simulate single channel communication via VHF3.
Rewrote ATSU/ATC pages to make use of the ATSU manager.

The datalink transmission time supports three different types:
- Instant: Transmission of data takes up to two seconds
- Fast: Transmission of data takes up to twenty seconds
- Real: Transmission of data takes up to 120 seconds (simulate CPDLC 240 roundtrip-time)

This feature can only be used if the pilot flies on VATSIM or IVAO.

Deactivated all additional pages of "ATC MENU" that are currently not supported by the datalink.
Will be reactivated in later PRs.

**This PR depends on the extension of the NXApi and the corresponding PR: https://github.com/flybywiresim/api/pull/119**
### AOC update of "Depart Req"
Updated AOC menu

### PDC request
grafik

### ATC Menu (MSN7560)
grafik

### ATC Notification
grafik
grafik

### Connection status
grafik
grafik

### MSG Record
grafik

### ATC MSG Display
grafik

### Full ATC command process on DCDU
grafik
grafik
grafik

### ATC MSG button
grafik

### EFB ATSU/AOC
grafik

### EFB Realism
grafik

## References

A320 FCOM documentation and used "MSN 7560" as a reference.
https://www.youtube.com/watch?v=KOdHYtF28Ts

## Additional context

The PR implements the http-based-communication protocol of hoppie.nl.
The API documentation can be found here:
http://www.hoppie.nl/acars/system/tech.html

After a short conversation with Jeroen (hoppie.nl) about missing CORS-headers is it required that the system uses a proxy to extend the HTTP headers of the server's response. This can be done by 'cors-anywhere'.

Discord username (if different from GitHub): sven

## Testing instructions

- Create a Hoppie-ID (http://www.hoppie.nl/acars/system/register.html)
- Activate before every flight Hoppe in the EFB (Settings/ATSU)
- 1. option: Fly online on VATSIM or IVAO and login to online stations (ATC/Communication/Notification)
- 2. option: Download http://www.hoppie.nl/acars/prg/atc/ and be your own controller
   - Start ACARS ATC
   - Logon as a station (i.e. TEST)
   - Logon on the station (ATC/Communication/Notification)

- Links to check if the communication is working well
   - List of all online stations and aircrafts: http://www.hoppie.nl/acars/system/online.html
   - List of all sent and received messages for a specific callsign: http://www.hoppie.nl/acars/system/callsign.html?network=IVAO&callsign=CALLSIGN
   
Also check the initial draft documentation:
https://docs-git-fork-frankkopp-featguidehoppie-flybywire.vercel.app/fbw-a32nx/feature-guides/hoppie/

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Comment by NUTTTTTTTT** [comment]
If this is still a work in progress, please mark as draft for now and mark ready for review once it's in a state to be QA tested and implemented. :)

**Comment by frankkopp** [comment]
When you write test instructions please also include some documentation on how to use it. 
I can then create a documentation page from it. 
Thanks

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_DepartReq.js
Comment (Benjozork) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_DepartReq.js:163]: **issue:** typo (addional instead of additional)

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/ATSU/A320_Neo_CDU_ATC_DepartReq.js
Comment (Benjozork) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/ATSU/A320_Neo_CDU_ATC_DepartReq.js:62]: **suggestion:** I'd also check for `getOrigin` not being null here

File: src/atsu/src/AocSystem.ts
Comment (Benjozork) [inline src/atsu/src/AocSystem.ts:29]:
```suggestion
    private connector: HoppieConnector | undefined = undefined;

    private messageQueue: AtsuMessage[] = [];
```

File: src/atsu/src/messages/AtsuMessage.ts
Comment (Benjozork) [inline src/atsu/src/messages/AtsuMessage.ts:72]:
```suggestion
    public Network = AtsuMessageNetwork.Hoppie;

    public UniqueMessageID: number | undefined = undefined;

    public Timestamp: AtsuTimestamp | undefined = undefined;

    public DcduTimestamp: number | undefined = undefined;

    public Station = '';

    public ComStatus: AtsuMessageComStatus = AtsuMessageComStatus.Open;

    public Type: AtsuMessageType | undefined = undefined;

    public Direction: AtsuMessageDirection | undefined = undefined;

    public Status: AtsuMessageResponseStatus | undefined = undefined;
```

File: src/atsu/src/messages/AtsuTimestamp.ts
Comment (Benjozork) [inline src/atsu/src/messages/AtsuTimestamp.ts:14]:
```suggestion
    public Year: number = SimVar.GetSimVarValue('E:ZULU YEAR', 'number');

    public Month: number = SimVar.GetSimVarValue('E:ZULU MONTH OF YEAR', 'number');

    public Day: number = SimVar.GetSimVarValue('E:ZULU DAY OF MONTH', 'number');

    public Seconds: number = SimVar.GetSimVarValue('E:ZULU TIME', 'seconds');
```

File: src/atsu/src/messages/CpdlcMessage.ts
Comment (Benjozork) [inline src/atsu/src/messages/CpdlcMessage.ts:20]:
```suggestion
    public Response: CpdlcMessageResponseType = undefined;
```

File: src/atsu/src/messages/FreetextMessage.ts
Comment (Benjozork) [inline src/atsu/src/messages/FreetextMessage.ts:10]:
```suggestion
    public Lines: string[] = [];
```

File: src/instruments/src/DCDU/elements/Button.tsx
Comment (Benjozork) [inline src/instruments/src/DCDU/elements/Button.tsx:8]:
```suggestion
    messageId: number,
    index: string,
    content: string,
    clicked: (btn: string) => void
```

File: src/instruments/src/DCDU/elements/Button.tsx
Comment (Benjozork) [inline src/instruments/src/DCDU/elements/Button.tsx:11]: **issue:** fix lint warning

**Comment by Benjozork** [comment]
Preliminary review requested by author

**Comment by svengcz** [comment]
@beheh: I saw that you retriggered the PR action.
I reviewed the findings about the code formatting, but cannot find the issues.
These issues in A32NX_ATSU.js were also found in the first run. I fixed it and pushed the changes but it looks like the action used the old commit for the test. Do I need to do something else?

**Comment by beheh** [comment]
@svengcz Hi! It's simply a GitHub protection against malicious PRs, so one of the maintainers has to approve that CI should be run across your commits. This should only be happening as this is your first PR though, and in future it should work as usual.

I've told GitHub to trigger the build again.

**Comment by Benjozork** [comment]
Not to dev-team: we are proceeding with reviews even though the fonts are not ready yet - we'll integrate them when all of them are finished

**Review by Benjozork — CHANGES_REQUESTED** [review]
Amazing work!

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/ATSU/A320_Neo_CDU_ATC_ConnectionNotification.js
Comment (svengcz) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/ATSU/A320_Neo_CDU_ATC_ConnectionNotification.js:4]: because I check on LSK2 and RSK2 in asynchronous functions if the page is still active.
if I don't do this and leave the page, the code returns after the asynchronous call is finished

File: src/atsu/src/AtcSystem.ts
Comment (svengcz) [reply src/atsu/src/AtcSystem.ts:37]: that is an hard bug... but that never triggered for me.
so I think I run into the raice condition every time
fixed it.

File: src/atsu/src/com/HoppieConnector.ts
Comment (svengcz) [reply src/atsu/src/com/HoppieConnector.ts:147]: Yes, this can also work. The IRL simulation is done in Datalink.js.
This queues (virtually) all send and receive requests and handles them step by step.
So the IRL simulation is not part of the poll-function

File: src/atsu/src/messages/PdcMessage.ts
Comment (svengcz) [reply src/atsu/src/messages/PdcMessage.ts:81]: you are right. fixed it

File: src/shared/src/nxapi.ts
Comment (svengcz) [reply src/shared/src/nxapi.ts:3]: didn't know that this library exists. will create a PR and remove this nxapi-stuff

File: src/instruments/src/DCDU/elements/Button.tsx
Comment (svengcz) [reply src/instruments/src/DCDU/elements/Button.tsx:9]: fixed

File: src/atsu/src/com/HoppieConnector.ts
Comment (svengcz) [reply src/atsu/src/com/HoppieConnector.ts:147]: updated to 5000

File: src/atsu/src/AtsuManager.ts
Comment (svengcz) [reply src/atsu/src/AtsuManager.ts:91]: changed to public properties

File: src/shared/src/nxapi.ts
Comment (svengcz) [reply src/shared/src/nxapi.ts:3]: fixed it and also removed NXApi from the MCDU main page.
centralized in AOC of the ATSU

**Review by Benjozork — COMMENTED** [review]
**question:** I see the line [here](https://github.com/flybywiresim/a32nx/blob/c984f5db1d0da8905d0afaafe97d05161962643b/flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js#L247) (in master) to update the telex (sends position data to map if consented) is no longer there. Did you move it somewhere?

**Comment by svengcz** [comment]
> 

Yes, it is now part of the [ATSU](https://github.com/svengcz/a32nx/blob/61441f8980b04e3dcec8d32e970a707eae1eed99/src/atsu/src/com/NXApiConnector.ts#L156).

And this poll-function is triggered [here](https://github.com/svengcz/a32nx/blob/61441f8980b04e3dcec8d32e970a707eae1eed99/src/atsu/src/com/Datalink.ts#L71).

The idea is to have a central location where all the ATSU-related communication is done

**Review by Benjozork — APPROVED** [review]
LGTM!

We can begin testing while fonts are being finished.

**Comment by Saschl** [comment]
Needs a fix in api-client, I'll remove the label once the fix is merged

**Comment by mico975** [comment]
Quality Assurance Tester Report

Discord          : mico#3145
Object of testing: #6451 
Tier of Testing  : 1
Date             : report on 03/02/2022

Testing Process: 
Lenghty discussion in the QA thread on discord. Basically tested all features multiple times.

Negatives:       
N/A

Testing Results: 
Passed

Conclusions: 
Nice job

Media:       
image
image

**Comment by RichardPilbery** [comment]
Quality Assurance  Report

Discord          : tricky_dicky[Z+0]3571
Object of testing: #6451 
Tier of Testing  : 1
Date             : 02/02/2022

Testing Process: 
Spawned C&D at EGLL
Imported Simbrief flight plan for EGLL-EDDH
Toggled Hoppie on EFB
Requested PDC from Heathrow 
Checked Hoppie website for confirmation that message was sent.
Received clearance.
Able to silence message notification with ATC MSG button and on ackwoleding messages on DCDU
Able to print out PDC
En route - Attemped to contact EHAW. Messages being passed to Hoppie as confirmed on website but controller could not receive (unlikely to be this PR)
No further Hoppie controllers available.

Negatives:       
N/A

Testing Results: 
Passed

Conclusions: 
PR works as intended. We are going to need to managed expectations of the Hoppie network as it appears to be a bit flaky. In addition, there are issues with controllers not logging off so a region appears available when it is not etc. etc.
Did find the need to re-activate Hoppie on the EFB every time I loaded in a bit annoying, but understand that this is to reduce the load on the Hoppie network.

Media:       
PDC-3
PDC-2

File: src/instruments/src/EFB/Settings/Settings.tsx
Comment (tracernz) [inline src/instruments/src/EFB/Settings/Settings.tsx:521]: This is not a setting. It should be a Connect/Disconnect button, not in the settings pages. Probably fits best on the ATC page.

**Review by tracernz — CHANGES_REQUESTED** [review]

File: src/instruments/src/EFB/Settings/Settings.tsx
Comment (svengcz) [reply src/instruments/src/EFB/Settings/Settings.tsx:521]: Moved the connect-stuff to the ATC page.

grafik

**Review by tracernz — APPROVED** [review]

**Comment by davidnorfolk96** [comment]
Quality Assurance Report

Discord: !davidn123#0001
Object of testing: 6451
Tier of Testing: 1
Date: 05/02/2022

Testing Process: Loaded up cold & dark at EGKK, Set up a TEST acars and tested different CPDLC comms. This included Unable, standbys & wilcos. All worked well. I then tested different messages such as "Climb and maintain" & " Don't go any faster than 300knots" etc. Connecting worked well & found no issues with this. I followed the steps as follows.... 

Activate before every flight Hoppe in the EFB (Settings/ATSU)-PASSED
option: Fly online on VATSIM or IVAO and login to online stations (ATC/Communication/Notification)-PASSED	
option: Download http://www.hoppie.nl/acars/prg/atc/ and be your own controller-PASSED
Start ACARS ATC-PASSED
Logon as a station (i.e. TEST)-PASSED
Logon on the station (ATC/Communication/Notification)-PASSED

Everything worked as intended. 




Negatives: N/A
Testing Results: Passed

Conclusions: I did find the CPDLC alarm abit low in quality and doesn't sound as crisp as it does in real life. However this is a minor thing. 
Media:
Screenshot_1
Screenshot_2

**Comment by NUTTTTTTTT** [comment]
Quality Assurance Tester/Trainee Report

Discord : █▀█ █▄█ ▀█▀#2123
Object of testing: (#PR 6451)
Tier of Testing : 1
Date : (06/02/2022)

Testing Process: 
Tested off network using Hoppie ATC client (http://www.hoppie.nl/acars/prg/atc/ ).
- Set up and launched ATC client  with unique station.
- Enabled Hoppie using personal log on code and tested connection via ATC page of the EFB - PASS.
- ATC > Notify tested and connection established to my ATC station - PASS.
- PDC requested, PDC message delivered to ATC client in full - PASS.
- Tested numerous ATC to Pilot messages and the available responses with my own station, as well as with Mico & Holland in QA thread, worked perfectly - PASS.
- ATC MSG rings as expected, light is cleared and ringtone stops when reading/actioning message via DCDU. Also deactivates when clicking ATC MSG - PASS.
- Checked the other functions not yet implemented are greyed/disabled on ATC page - PASS.

Negatives:   N/A  

Testing Results: Passed

Conclusions: Great PR for a long awaited feature, haven't been in the A32NX properly in a while but had a blast testing this at length and getting any quirks ironed out! Great stuff Sven!

Media:       
image
image
image
image
image
image
image

**Comment by NUTTTTTTTT** [comment]
Retested a third time to test the updated font commit, all good and PASS remains! understand the new font will be pushed in a later PR once finalised!

image
image
image
image

**Comment by NUTTTTTTTT** [comment]
Retested with dev proxy removed - PASS.

image

**Pull request merged**
